### PR TITLE
demo playback QOL features

### PR DESCRIPTION
**Pull request by rfomin**

### CODE REVIEW

File: Source/i_system.c [hunk 1]
```diff
@@ -219,21 +219,22 @@ void I_InitKeyboard(void)
...
-void I_Init(void)
+void I_InitTimer(void)
 {
    int p;
 
+   int time_ms = SDL_GetTicks();
+   int old_time = time_ms * clock_rate / 100;
+   int delta = (time_ms * realtic_clock_rate / 100) - old_time;
+
+   basetime += delta;
```
Comment (rfomin) [inline Source/i_system.c:230]: @fabiangreffrath I'm struggling a little here. I think we need to change the `basetime` to compensate for the speed change, but that doesn't work right.

Comment (fabiangreffrath) [reply Source/i_system.c:230]: Hm, I am not sure if we even need gradual increase/decrease of game speed. In Crispy (HEAD), for example, I have a key to merely switch between playdemo/timedemo behaviour, and this comes with a lot less hassle, obviously.

File: Source/hu_stuff.c [hunk 2]
```diff
@@ -969,6 +969,27 @@ static void HU_DrawCrosshair(void)
...
+boolean HU_DemoProgressBar(boolean force)
+{
+  const int progress = SCREENWIDTH * defdemotics / deftotaldemotics;
+  static int old_progress = 0;
+
+  if (progress - old_progress)
+  {
+    old_progress = progress;
+  }
+  else if (!force)
```
[the remark this answers is not in the document]
Comment (rfomin) [reply Source/hu_stuff.c:982]: I like to draw progress bar even if `nodrawers == true`, it's useful for `demowarp`. See beginning of `I_Display`.

File: Source/i_system.c (hunk 1 above)
Comment (rfomin) [reply Source/i_system.c:230]: We have this cool game speed feature, I want to use it more. Okay, I'll try again, and if it doesn't work, I'll revert it.

**Comment by fabiangreffrath** [comment]
What exactly is the difference between demoskip and demonext?

**Comment by rfomin** [comment]
> What exactly is the difference between demoskip and demonext?

demonext goes to the next level, demoskip stops only at the user's command.

**Comment by fabiangreffrath** [comment]
So, demoskip is like that playdemo/timedemo switch that I have in Crispy?

**Comment by rfomin** [comment]
> So, demoskip is like that playdemo/timedemo switch that I have in Crispy?

Yes, but it uses I_EnableWarp()

**Comment by fabiangreffrath** [comment]
So, you don't even see anything?

**Comment by rfomin** [comment]
> So, you don't even see anything?

Just a progress bar. I copied this from PrBoom+. We also have keys to increase the speed.

**Comment by rfomin** [comment]
I think timedemo on a modern PC is too fast to see anything in most cases.

**Comment by fabiangreffrath** [comment]
But then I don't even have a clue when might be the right moment to un-skip again?

**Comment by rfomin** [comment]
> But then I don't even have a clue when might be the right moment to un-skip again?

Progress bar is a clue :) But ok, we can change this to timedemo.

**Comment by fabiangreffrath** [comment]
Yes, please.

**Comment by rfomin** [comment]
With vsync timedemo is too slow on a 60hz screen. So we need to disable vsync?

**Comment by rfomin** [comment]
Alternatively we can set `realtic_clock_rate = 1000` or fastdemo for the same effect.

**Comment by fabiangreffrath** [comment]
Okay, could we go the fastdemo route?

File: Source/i_system.c [hunk 3]
```diff
@@ -90,9 +95,24 @@ int I_GetTime_RealTime(void)
...
+
+  if (clock_rate == 100)
+    time = SDL_GetTicks();
+  else
+    time = SDL_GetTicks() * clock_rate / 100;
+
+  if (basetime == 0)
+    basetime = time;
+
+  return time - basetime;
```
Comment (fabiangreffrath) [inline Source/i_system.c:110]: I believe the `clock_rate / 100` factor has to be applied to the `time - basetime` difference? Imagine we start with `clock_rate == 100` and then immediately switch to `clock_rate = 10` in the next tic. Then we would substract a `time` value that is smaller than `basetime`, or not?

File: Source/m_menu.c [hunk 4]
```diff
@@ -2821,7 +2825,8 @@ setup_menu_t keys_settings3[] =
   {"MISCELLANEOUS",S_SKIP|S_TITLE,m_null,KB_X,M_Y+11*M_SPC},
   {"RELOAD LEVEL",S_INPUT,m_scrn,KB_X,M_Y+12*M_SPC,{0},input_menu_reloadlevel},
   {"NEXT LEVEL"  ,S_INPUT,m_scrn,KB_X,M_Y+13*M_SPC,{0},input_menu_nextlevel},
-  {"FINISH DEMO" ,S_INPUT,m_scrn,KB_X,M_Y+14*M_SPC,{0},input_demo_quit},
+  {"SKIPPING DEMO",S_INPUT,m_scrn,KB_X,M_Y+14*M_SPC,{0},input_demo_skip},
```
Comment (fabiangreffrath) [inline Source/m_menu.c:2828]: Could we call this `fast-forward` or just `forward` or similar?

File: Source/m_menu.c [hunk 5]
```diff
@@ -5076,15 +5085,50 @@ boolean M_Responder (event_t* ev)
...
 			return true;
 	}
+
+        if (M_InputActivated(input_demo_skip))
+        {
+          if (demoplayback && singledemo && !demonext &&
+              demowarp == -1 && demoskip_tics == -1)
+          {
+            fastdemo = !fastdemo;
+            I_InitTimer();
```
Comment (fabiangreffrath) [inline Source/m_menu.c:5105]: `basetime` gets re-evaluated in `I_InitTimer()`. Is this intended?

File: Source/i_system.c (hunk 3 above)
Comment (rfomin) [reply Source/i_system.c:110]: `basetime` adjusted in `I_InitTimer()`

File: Source/m_menu.c (hunk 5 above)
Comment (rfomin) [reply Source/m_menu.c:5105]: Yes, it's intended for game speed change compensation.

**Comment by rfomin** [comment]
@fabiangreffrath What do you think about refactoring `i_system.c`? We could put the timer functions in `i_timer.c` and remove the dead code (keboard init).

File: Source/i_video.c [hunk 6]
```diff
@@ -896,7 +888,7 @@ int widescreen; // widescreen mode
 
 void I_FinishUpdate(void)
 {
-   if (noblit || !in_graphics_mode)
+   if (!in_graphics_mode)
       return;
```
Comment (fabiangreffrath) [inline Source/i_video.c:892]: I think the `noblit` mode may even prove useful. In contrast to `nodraw` which doesn't even call Doom's drawing functions, this only omits the rendering from framebuffer to screen. So, if you let `noblit` and `nodraw` run against each other, you'll find out how much time you spend in SDL's rendering part.

**Pull request merged**